OrbitGgp/include/OrbitGgp/Client.h

```diff
@@ -15,22 +16,24 @@
 
 namespace OrbitGgp {
 
-class Client {
+class Client : public QObject {
+  Q_OBJECT
+
  public:
-  static outcome::result<Client> Create();
+  static QPointer<Client> Create(QObject* parent);
 
-  int GetNumberOfRequestsRunning() const { return number_of_requests_running_; }
+  void GetInstances();
+  void GetSshInfo(const Instance& ggp_instance);
 
-  void GetInstancesAsync(
-      const std::function<void(outcome::result<QVector<Instance>>)>& callback);
-  void GetSshInformationAsync(
-      const Instance& ggpInstance,
-      const std::function<void(outcome::result<SshInfo>)>& callback);
+ signals:
+  void instancesReturned(outcome::result<QVector<Instance>>);
+  void sshInfoReturned(outcome::result<SshInfo>);
```

Comment:
This feels a bit skewed from a design perspective. If we can make multiple (parallel) requests (that's how RunProcessWithTimeout is designed) then we also need a way to differentiate between the results coming back. But with these signals we can't. We have to create multiple clients to do so.

Alternatively RunProcessWithTimeout should not have this additional complexity with creating `process` and `timeout_timer` on the heap. These should be member variables of `Client`. And talking about `Client`: In this second case the name would be off, it's more a `Request` maybe?

Since we don't need the parallel thing right now, I'm fine with either design.

Reply:
I see your point. A good design might be abstracting a "command line with timeout class" that supports signals. I looked at quickly at turning the `Client` into `Request` and then having one signal that the request returned. To make that nice, I wanted to have a templated request return type, but that seems not easily possible, since `Q_OBJECT` does not support templates. 
So im going to roll this back to use callbacks instead of signals, to have theoretically well defined behaviour with parallel requests (even thought we don't use the currently)

Reply:
Yeah, Q_OBJECT and templates is kind of tricky. You could theoretically do this with a templated subclass. But I recommend not going down this road. Passing a typed-erased parameter type in the signal would be the Qt-way of doing this. QVariant for example.

Reply:
Looking at `QVariant`, it can not hold custom types out of the box, right? example: `QVariant{OrbitGgp::SshInfo()}` would not work unless I change `SshInfo`.

Reply:
Have a look at QVariant::fromValue: https://doc.qt.io/qt-5/qvariant.html#fromValue
You need to register the custom type. Then it should work.